tests/MongoDB.Driver.Core.Tests/Specifications/connection-monitoring-and-pooling/ConnectionMonitoringAndPoolingTestRunner.cs

```diff
@@ -0,0 +1,609 @@
+                    return connection;
+                });
+            var connectionPool = new ExclusiveConnectionPool(
+                serverId,
+                endPoint,
+                connectionPoolSettings,
+                connectionFactory.Object,
+                eventSubscriber);
+
+            return connectionPool;
+        }
+
+        private void Start(BsonDocument operation, ConcurrentDictionary<string, Task> tasks)
+        {
+            var startTarget = operation.GetValue("target").ToString();
+            tasks.GetOrAdd(startTarget, (Task)null);
+        }
+
+        private void WaitForEvent(EventCapturer eventCapturer, BsonDocument operation)
+        {
+            var eventType = MapExpectedEventName(operation.GetValue("event").ToString());
+            if (eventType == null)
+            {
+                return;
+            }
+
+            var expectedCount = operation.GetValue("count").ToInt32();
+            var notifier = new EventNotifier(
+                eventCapturer,
+                coll => coll.Count(c => c.GetType().Name == eventType) >= expectedCount);
+
+            var testFailedTimeout = Task.Delay(TimeSpan.FromMinutes(5), CancellationToken.None);
+            var index = Task.WaitAny(notifier.TaskCompletion, testFailedTimeout);
+            if (index != 0)
+            {
+                throw new Exception($"{nameof(WaitForEvent)} executing is too long.");
+            }
+        }
+
+        private void WaitForThread(BsonDocument operation, ConcurrentDictionary<string, Task> tasks, out Exception exception)
+        {
+            exception = null;
+            var waitThread = operation.GetValue("target").ToString();
+            if (tasks.TryGetValue(waitThread, out var task) && task != null)
+            {
+                exception = Record.Exception(() => task.GetAwaiter().GetResult());
+            }
+            else
+            {
+                throw new Exception($"The task {waitThread} must be configured before waiting.");
+            }
+        }
+
+        // nested types
+        private class TestCaseFactory : JsonDrivenTestCaseFactory
+        {
+            protected override string PathPrefix => "MongoDB.Driver.Core.Tests.Specifications.connection_monitoring_and_pooling.tests.";
+
+            protected override IEnumerable<JsonDrivenTestCase> CreateTestCases(BsonDocument document)
+            {
+                var version = document.GetValue("version", null)?.ToInt32() ?? 1;
+                foreach (var async in new[] { false, true })
+                {
+                    var name = GetTestCaseName(document, document, version);
+                    name = $"{name}:async={async}";
+                    var test = document.DeepClone().AsBsonDocument.Add("async", async);
+                    yield return new JsonDrivenTestCase(name, test, test);
+                }
+            }
+        }
+
+        private class ThreadPerTaskScheduler : TaskScheduler
```

Comment:
this code is copied from that place: https://code.msdn.microsoft.com/Samples-for-Parallel-b4b76364/sourcecode?fileId=44488&pathId=2098696067

As I see from the description and my local tests, this is exactly what we want to see in this test. With this `taskScheduler` each task will be launched in a new thread.